### PR TITLE
bpo-46753: Add the empty tuple to the _PyRuntimeState.global_objects.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
https://bugs.python.org/issue46753

### CODE REVIEW

File: Objects/tupleobject.c [hunk 1]
```diff
@@ -265,40 +222,33 @@ PyTuple_Pack(Py_ssize_t n, ...)
 static void
 tupledealloc(PyTupleObject *op)
 {
-    Py_ssize_t len =  Py_SIZE(op);
+    if ((PyObject *)op == tuple_get_empty()) {
```
Comment (markshannon) [inline Objects/tupleobject.c]: Having the function call here is a bit odd. `op == &_Py_SINGLETON(tuple_empty)`

File: Objects/tupleobject.c [hunk 2]
```diff
@@ -100,58 +95,20 @@ tuple_alloc(Py_ssize_t size)
...
-#ifdef Py_DEBUG
-    assert(state->numfree[0] != -1);
-#endif
-
-    Py_INCREF(op);
-    return (PyObject *) op;
-#else
-    return PyTuple_New(0);
-#endif
+    return (PyObject *)&_Py_SINGLETON(tuple_empty);
```
Comment (markshannon) [inline Objects/tupleobject.c:66]: Don't you need an incref here?

File: Objects/tupleobject.c [hunk 3]
```diff
@@ -57,27 +57,22 @@ static PyTupleObject *
...
 {
     PyTupleObject *op;
-#if PyTuple_MAXSAVESIZE > 0
-    // If Python is built with the empty tuple singleton,
-    // tuple_alloc(0) must not be called.
-    assert(size != 0);
-#endif
-    if (size < 0) {
+    /* The empty tuple is statically allocated. */
+    if (size <= 0) {
```
Comment (kumaraditya303) [inline Objects/tupleobject.c]: Can this be an assert as it is in the middle of a hot path?

File: Objects/tupleobject.c [hunk 4]
```diff
@@ -100,58 +95,20 @@ tuple_alloc(Py_ssize_t size)
...
-#ifdef Py_DEBUG
-    assert(state->numfree[0] != -1);
-#endif
-
-    Py_INCREF(op);
-    return (PyObject *) op;
-#else
-    return PyTuple_New(0);
-#endif
+    return _Py_NewRef((PyObject *)&_Py_SINGLETON(tuple_empty));
```
Comment (kumaraditya303) [inline Objects/tupleobject.c]:
```suggestion
    return Py_NewRef((PyObject *)&_Py_SINGLETON(tuple_empty));
```

**Comment by ericsnowcurrently** [comment]
Note that https://github.com/python/cpython/pull/31345/commits/c163aca6973ac06611daefd4966089b8337a8a07 is the actual global empty tuple part.  Everything else is cleanup (in particularly to make sure I got the freelist stuff right).

**Pull request merged**